File: packages/types/src/callbacks.ts
/**
 * Defines callback interfaces for inference and agent interactions.
 * Imports: Utilizes types from `history.js`, `inference.js`, `stats.js`, and `tools.js`.
 * @example
 * // Example of using InferenceCallbacks
 * const inferenceCallbacks: InferenceCallbacks = {
 *   onToken: (token: string, from: string) => console.log(`Token from ${from}: ${token}`),
 *   onError: (err: any, from: string) => console.error(`Error from ${from}:`, err)
 * };
 *
 * // Example of using AgentCallbacks
 * const agentCallbacks: AgentCallbacks = {
 *   onToolCall: (tc: ToolCallSpec, type: string, from: string) => console.log(tc),
 *   onTurnEnd: (ht: HistoryTurn, from: string) => console.log(ht)
 * };
 *
 * // Example of using AllCallbacks
 * const allCallbacks: AllCallbacks = {
 *   ...inferenceCallbacks,
 *   ...agentCallbacks
 * };
 */

import type { HistoryTurn, ToolTurn } from "./history.js";
import type { InferenceResult } from "./inference.js";
import type { PromptProcessingInProgressStats } from "./stats.js";
import type { ToolCallSpec } from "./tools.js";

/**
 * Represents the callbacks for inference events.
 *
 * @interface InferenceCallbacks
 * @property {() => void} [onStartThinking] - Callback when thinking starts.
 * @property {() => void} [onEndThinking] - Callback when thinking ends.
 * @property {(t: string, from: string) => void} [onToken] - Callback for each token emitted.
 * @property {(t: string, from: string) => void} [onThinkingToken] - Callback for thinking tokens.
 * @property {(t: string, from: string) => void} [onToolCallToken] - Callback for tool call tokens.
 * @property {(data: PromptProcessingInProgressStats, from: string) => void} [onStartEmit] - Callback when emission starts.
 * @property {(result: InferenceResult, from: string) => void} [onEndEmit] - Callback when emission ends.
 * @property {(err: any, from: string) => void} [onError] - Callback for errors.
 * @property {(tc: Array<ToolCallSpec>, from: string) => void} [onToolCallInProgress] - Callback for tool call progress.
 * @property {(progress: PromptProcessingInProgressStats, from: string) => void} [onPromptProcessingProgress] - Callback for prompt processing progress.
 * @example
 * const inferenceCallbacks: InferenceCallbacks = {
 *   onToken: (token: string, from: string) => console.log(`Token from ${from}: ${token}`),
 *   onError: (err: any, from: string) => console.error(`Error from ${from}:`, err),
 *   onStartThinking: (from: string) => console.log(`Thinking started in ${from}`),
 *   onEndThinking: (from: string) => console.log(`Thinking ended in ${from}`)
 * };
 */
interface InferenceCallbacks {
    onStartThinking?: (from: string) => void;
    onEndThinking?: (from: string) => void;
    onToken?: (t: string, from: string) => void;
    onThinkingToken?: (t: string, from: string) => void;
    onToolCallToken?: (t: string, from: string) => void;
    onStartEmit?: (data: PromptProcessingInProgressStats, from: string) => void;
    onEndEmit?: (result: InferenceResult, from: string) => void;
    onError?: (err: any, from: string) => void;
    onToolCallInProgress?: (tc: Array<ToolCallSpec>, from: string) => void;
    onPromptProcessingProgress?: (progress: PromptProcessingInProgressStats, from: string) => void;
}

/**
 * Represents the callbacks for agent interactions.
 *
 * @interface AgentCallbacks
 * @property {(tc: ToolCallSpec, type: string, from: string) => void} [onToolCall] - Callback for tool calls.
 * @property {(tc: ToolCallSpec, tr: any, type: string, from: string) => void} [onToolCallEnd] - Callback for tool call completion.
 * @property {(tc: Array<ToolCallSpec>, from: string) => void} [onToolsTurnStart] - Callback for tools turn start.
 * @property {(tt: Array<ToolTurn>, from: string) => void} [onToolsTurnEnd] - Callback for tools turn end.
 * @property {(from: string) => void} [onTurnStart] - Callback when a new turn starts.
 * @property {(ht: HistoryTurn, from: string) => void} [onTurnEnd] - Callback for turn end.
 * @property {(txt: string, from: string) => void} [onAssistant] - Callback for assistant text.
 * @property {(txt: string, from: string) => void} [onThink] - Callback for thinking text.
 * @example
 * const agentCallbacks: AgentCallbacks = {
 *   onToolCall: (tc: ToolCallSpec, type: string, from: string) => console.log(`Tool call from ${from}:`, tc),
 *   onTurnEnd: (ht: HistoryTurn, from: string) => console.log(`Turn ended in ${from}`),
 *   onAssistant: (txt: string, from: string) => console.log(`Assistant message from ${from}:`, txt)
 * };
 */
interface AgentCallbacks {
    onToolCall?: (tc: ToolCallSpec, type: string, from: string) => void;
    onToolCallEnd?: (tc: ToolCallSpec, tr: any, type: string, from: string) => void;
    onToolsTurnStart?: (tc: Array<ToolCallSpec>, from: string) => void;
    onToolsTurnEnd?: (tt: Array<ToolTurn>, from: string) => void;
    onTurnStart?: (from: string) => void;
    onTurnEnd?: (ht: HistoryTurn, from: string) => void;
    onAssistant?: (txt: string, from: string) => void;
    onThink?: (txt: string, from: string) => void;
}

/**
 * Combines inference and agent callbacks into a single interface.
 *
 * @interface AllCallbacks
 * @extends InferenceCallbacks
 * @extends AgentCallbacks
 * @example
 * const allCallbacks: AllCallbacks = {
 *   onToken: (token: string, from: string) => console.log(`Token: ${token}`),
 *   onToolCall: (tc: ToolCallSpec, type: string, from: string) => console.log(`Tool: ${tc}`),
 *   onError: (err: any, from: string) => console.error(`Error:`, err)
 * };
 */
interface AllCallbacks extends InferenceCallbacks, AgentCallbacks { }

export {
    InferenceCallbacks,
    AgentCallbacks,
    AllCallbacks,
}

File: server/go/callbacks/callbacks.go
package callbacks

import (
	"crypto/rand"
	"encoding/json"
	"fmt"
	"io"
	"sync"

	"github.com/synw/agent-smith/server/go/state"
	"github.com/synw/agent-smith/server/go/types"
	"github.com/synw/agent-smith/server/go/utils"
	"github.com/synw/agent-smith/server/go/websock"
)

// CallbackHandlers bridges execution events to WebSocket messages.
type CallbackHandlers struct {
	ws               websock.WSConn
	from             string
	confirmToolCalls map[string]*utils.Awaiter
	mu               sync.Mutex // protects confirmToolCalls map
}

// NewCallbackHandlers creates a new CallbackHandlers instance.
func NewCallbackHandlers(ws websock.WSConn, from string) *CallbackHandlers {
	return &CallbackHandlers{
		ws:               ws,
		from:             from,
		confirmToolCalls: make(map[string]*utils.Awaiter),
	}
}

// sendMsg sends a WsRawServerMsg over the WebSocket connection.
func (cb *CallbackHandlers) sendMsg(msgType types.WsServerMsgType, msg string) {
	rsm := types.WsRawServerMsg{
		Type: msgType,
		From: cb.from,
		Msg:  msg,
	}
	data, err := json.Marshal(rsm)
	if err != nil {
		if state.IsDebug.Load() {
			fmt.Printf("Failed to marshal message: %v\n", err)
		}
		return
	}
	if err := cb.ws.Send(data); err != nil {
		if state.IsDebug.Load() {
			fmt.Printf("Failed to send message: %v\n", err)
		}
	}
}

// BuildOptions returns a map of callback functions for the lm binary execution.
func (cb *CallbackHandlers) BuildOptions() map[string]interface{} {
	return map[string]interface{}{
		"onStartEmit": func(p types.PromptProcessingInProgressStats, from string) {
			data, err := json.Marshal(p)
			if err != nil {
				fmt.Printf("Failed to marshal start emit: %v\n", err)
				return
			}
			cb.sendMsg(types.StartEmitType, string(data))
		},
		"onToken": func(t string, from string) {
			cb.sendMsg(types.TokenType, t)
		},
		"onThinkingToken": func(t string, from string) {
			cb.sendMsg(types.ThinkingTokenType, t)
		},
		"onStartThinking": func(from string) {
			cb.sendMsg(types.ThinkingStartType, "")
		},
		"onEndThinking": func(from string) {
			cb.sendMsg(types.ThinkingEndType, "")
		},
		"onTurnStart": func(from string) {
			cb.sendMsg(types.TurnStartType, "")
		},
		"onTurnEnd": func(ht map[string]interface{}, from string) {
			data, err := json.Marshal(ht)
			if err != nil {
				fmt.Printf("Failed to marshal turn end: %v\n", err)
				return
			}
			cb.sendMsg(types.TurnEndType, string(data))
		},
		"onAssistant": func(txt string, from string) {
			cb.sendMsg(types.AssistantType, txt)
		},
		"onThink": func(txt string, from string) {
			cb.sendMsg(types.ThinkType, txt)
		},
		"onEndEmit": func(res types.InferenceResult, from string) {
			data, err := json.Marshal(res)
			if err != nil {
				fmt.Printf("Failed to marshal end emit: %v\n", err)
				return
			}
			cb.sendMsg(types.EndEmitType, string(data))
		},
		"onToolCallToken": func(t string, from string) {
			data, err := json.Marshal(t)
			if err != nil {
				fmt.Printf("Failed to marshal tool call token: %v\n", err)
				return
			}
			cb.sendMsg(types.ToolCallTokenType, string(data))
		},
		"onToolCallInProgress": func(tcs []interface{}, from string) {
			data, err := json.Marshal(tcs)
			if err != nil {
				fmt.Printf("Failed to marshal tool call in progress: %v\n", err)
				return
			}
			cb.sendMsg(types.ToolCallInProgressType, string(data))
		},
		"onPromptProcessingProgress": func(progress types.PromptProcessingInProgressStats, from string) {
			data, err := json.Marshal(progress)
			if err != nil {
				fmt.Printf("Failed to marshal prompt processing progress: %v\n", err)
				return
			}
			cb.sendMsg(types.PromptProcessingProgress, string(data))
		},
		"onToolsTurnStart": func(tcs map[string]interface{}, from string) {
			data, err := json.Marshal(tcs)
			if err != nil {
				fmt.Printf("Failed to marshal tools turn start: %v\n", err)
				return
			}
			cb.sendMsg(types.ToolsTurnStartType, string(data))
		},
		"onToolsTurnEnd": func(tr map[string]interface{}, from string) {
			data, err := json.Marshal(tr)
			if err != nil {
				fmt.Printf("Failed to marshal tools turn end: %v\n", err)
				return
			}
			cb.sendMsg(types.ToolsTurnEndType, string(data))
		},
		"onToolCall": func(tc map[string]interface{}, typeStr string, from string) {
			if tcID, ok := tc["id"].(string); !ok || tcID == "" {
				tc["id"] = generateUUID()
			}
			payload := map[string]interface{}{
				"tc":   tc,
				"type": typeStr,
				"from": from,
			}
			data, err := json.Marshal(payload)
			if err != nil {
				fmt.Printf("Failed to marshal tool call: %v\n", err)
				return
			}
			cb.sendMsg(types.ToolCallType, string(data))
		},
		"onToolCallEnd": func(tc map[string]interface{}, tr interface{}, typeStr string, from string) {
			var toolResData string
			if trObj, ok := tr.(map[string]interface{}); ok {
				trObj["type"] = typeStr
				if text, exists := trObj["text"]; exists {
					toolResData = fmt.Sprintf("%v", text)
				} else {
					data, _ := json.Marshal(trObj)
					toolResData = string(data)
				}
			} else if trStr, ok := tr.(string); ok {
				toolResData = trStr
			} else {
				toolResData = fmt.Sprintf("%v", tr)
			}
			payload := map[string]interface{}{
				"tc":   tc,
				"type": typeStr,
				"from": from,
			}
			payloadData, err := json.Marshal(payload)
			if err != nil {
				fmt.Printf("Failed to marshal tool call end payload: %v\n", err)
				return
			}
			toolCallEndMsg := string(payloadData) + "<|xtool_call_id|>" + toolResData
			cb.sendMsg(types.ToolCallEndType, toolCallEndMsg)
		},
		"onConfirmToolUsage": func(tc map[string]interface{}, from string) (bool, error) {
			if tcID, ok := tc["id"].(string); !ok || tcID == "" {
				tc["id"] = generateUUID()
			}
			tcData, err := json.Marshal(tc)
			if err != nil {
				fmt.Printf("Failed to marshal tool confirmation: %v\n", err)
				return false, err
			}
			cb.sendMsg(types.ToolCallConfirmType, string(tcData))

			awaiter := utils.CreateAwaiter()
			tcID := tc["id"].(string)
			cb.mu.Lock()
			cb.confirmToolCalls[tcID] = awaiter
			cb.mu.Unlock()

			result := awaiter.Wait()
			cb.mu.Lock()
			delete(cb.confirmToolCalls, tcID)
			cb.mu.Unlock()
			return result, nil
		},
		"nocli": true,
	}
}

// SendToken sends a token message.
func (cb *CallbackHandlers) SendToken(token string) {
	cb.sendMsg(types.TokenType, token)
}

// SendThinkingToken sends a thinking token message.
func (cb *CallbackHandlers) SendThinkingToken(token string) {
	cb.sendMsg(types.ThinkingTokenType, token)
}

// SendEndEmit sends an endemit message.
func (cb *CallbackHandlers) SendEndEmit(result types.InferenceResult) {
	data, err := json.Marshal(result)
	if err != nil {
		fmt.Printf("Failed to marshal end emit result: %v\n", err)
		return
	}
	cb.sendMsg(types.EndEmitType, string(data))
}

// SendFinalResult sends a finalresult message.
func (cb *CallbackHandlers) SendFinalResult(result types.InferenceResult) {
	data, err := json.Marshal(result)
	if err != nil {
		fmt.Printf("Failed to marshal final result: %v\n", err)
		return
	}
	cb.sendMsg(types.FinalResultType, string(data))
}

// From returns the source identifier.
func (cb *CallbackHandlers) From() string {
	return cb.from
}

// ResolveToolConfirmation resolves a pending tool confirmation awaiter.
func (cb *CallbackHandlers) ResolveToolConfirmation(id string, value bool) {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	if awaiter, exists := cb.confirmToolCalls[id]; exists {
		awaiter.Resolve(value)
		delete(cb.confirmToolCalls, id)
	}
}

// generateUUID generates a random UUID v4.
func generateUUID() string {
	uuid := make([]byte, 16)
	if _, err := io.ReadFull(rand.Reader, uuid); err != nil {
		return "00000000-0000-4000-8000-000000000000"
	}
	uuid[6] = (uuid[6] & 0x0f) | 0x40
	uuid[8] = (uuid[8] & 0x3f) | 0x80
	return fmt.Sprintf("%08x-%04x-%04x-%04x-%012x",
		uuid[0:4], uuid[4:6], uuid[6:8], uuid[8:10], uuid[10:16])
}

// toMap converts an interface{} to map[string]interface{}.
func toMap(v interface{}) map[string]interface{} {
	switch val := v.(type) {
	case map[string]interface{}:
		return val
	case map[string]string:
		m := make(map[string]interface{})
		for k, v := range val {
			m[k] = v
		}
		return m
	default:
		return nil
	}
}
